File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        int l = 0 ; int r=  h.size()-1;  int lmax = INT_MIN; int rmax = INT_MIN; 
        int total = 0; 
        while(l < r) {
             lmax = max(h[l], lmax); 
             rmax = max(h[r], rmax); 
             total += (lmax < rmax) ? lmax - h[l++] : rmax - h[r--]; 
        }
        return total ; 
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        l = 0
        r = len(h)-1

        lm =0
        rm = 0
        total = 0

        while l < r:
            hl = h[l]
            hr = h[r]
            if h[l] <= h[r]:
                lm = max(lm, hl)
                total += lm - hl
                l+=1
            else :
                rm = max(rm, hr)
                total+= rm - hr 
                r-=1
        return total 

            

            





        